### PR TITLE
Prevent index failed precondition

### PR DESCRIPTION
**Pull request by stuartnelson3**
this patch closes the `publishReady` channel only
if all preconditions have been met. Previously,
any result from the waitReady check would result
in unblocking the roundtripper that was supposed
to prevent indexing until all preconditions had
been met.

when the server is shut down, the ctx in
s.waitReady is canceled. this caused the function
to return and execute its defer statement, closing
`publishReady`.  this resulted in any queued events
being sent to the elasticsearch, which is not the
desired behavior.

closes https://github.com/elastic/apm-server/issues/8018

# how to test

1. start elasticsearch/kibana
2. navigate to fleet ui / agent policies; remove apm-1
3. go to add integrations / apm / settings; remove the integration
4. start apm server; verify preconditions aren't met from log output
5. send events to apm-server
6. stop apm-server
7. verify no events were indexed

### CODE REVIEW

File: beater/waitready.go [hunk 1]
```diff
@@ -79,10 +81,17 @@ func waitReady(
...
 }
 
 func (c *waitReadyRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
 	select {
+	case <-c.drain:
+		return &http.Response{
+			Request:    r,
+			StatusCode: http.StatusPreconditionFailed,
+			Body:       io.NopCloser(new(bytes.Buffer)),
+		}, nil
```
Comment (stuartnelson3) [inline beater/waitready.go:94]: choosing StatusPreconditionFailed was somewhat arbitrary; it stops the es client from retrying (I originally was returning an error, but then the client would retry X times) and has the word "precondition". Happy to change this to something else.

Comment (axw) [reply beater/waitready.go:94]: I don't think it matters a great deal, as this is all internal to the server, but I'd prefer not to use `http.StatusPreconditionFailed` as it's a client error (4xx). Perhaps use `http.StatusInternalServerError`?

I think ideally we would return an error. It's possible to customise the retry-on-error behaviour of go-elasticsearch, but it's probably a bit of a rigmarole just for this?

File: beater/waitready.go [hunk 2]
```diff
@@ -79,10 +81,17 @@ func waitReady(
...
+	drain <-chan struct{}
 }
 
 func (c *waitReadyRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
 	select {
+	case <-c.drain:
+		return &http.Response{
+			Request:    r,
+			StatusCode: http.StatusPreconditionFailed,
+			Body:       io.NopCloser(new(bytes.Buffer)),
```
Comment (axw) [inline beater/waitready.go:93]:
```suggestion
			Body:       io.NopCloser(strings.NewReader("server shutting down")),
```

might be a bit more understandable in logs?

**Review by axw — APPROVED** [review]
LGTM! Just a couple of minor things.

**Comment by stuartnelson3** [comment]
@axw surprisingly little rigamarole, actually. let me know if this looks good to you!

**Review by axw — APPROVED** [review]
Nice :) LGTM, thanks!

**Comment by axw** [comment]
Verified with 8.3.0 BC2. I ran the stack locally, and ran the apm-server image locally. Starting apm-server, sending some data, and stopping the server did not lead to any indices being created. OTOH, running with 8.2.0 did produce indices.

**Pull request merged**